doc/rtd/explanation/format.rst

```diff
@@ -139,6 +133,10 @@ The boothook is different in that:
   before any cloud-init modules are run.
 * It is run on every boot
 
+.. warning::
+    Use of ``INSTANCE_ID`` variable within boothooks is deprecated.
```

Comment:
Why do we want to deprecate this behavior? I get trying to drop the example as it is not a good suggestion for our users to dump artifacts into /var/lib/cloud/first-instance-id and we have a replacement for instance-id gating in `cloud-init-per instance` as you've said.

I don't see what cost there is to retain the environment variable $INSTANCE_ID for boot hooks to use if desired.  Also, this PR doesn't deprecate w/ deprecation message that really anywhere noticeable for a user. To deprecate that properly, it'd feels like it'd be 'expensive' in 
 cloudinit/handlers/boot_hook.py:handle_part  to actually grep for INSTANCE_ID in the boothook payload and log.deprecated if found.   We could do this if we think it's worth deprecating visbility to cloud-init's instance-id in boot hooks.

I mean, rather than environment variables, we also have the alternative `cloud-init query instance-id `  which provides the same feature. But, boothooks would have to invoke that on the cmdline as opposed to just being provided with that environment var.

Reply:
My earlier version was a bit too hasty, but I recently pushed a version that deprecates it in favor of jinja templating. Let me know if you think that is better.

I'm thinking of this more as a  "soft" deprecation. I have no plans to remove it anytime soon. Rather, we have better ways to accomplish the same thing (cloud-init-per or (as I just added) jinja templating) that will work consistently throughout user data rather than having magic variables that show up if a few random places. We could just leave it undocumented, but it seems better to me to have it documented as deprecated in case there's some good reason to remove in in the future.

> To deprecate this properly, it'd feels like it'd be 'expensive' 

I think this is good enough. Warnings rising from the code are nice to have, but I don't think they're must have. If we ever do need to break INSTANCE_ID, we could just list it as a breaking change and patch it out.

> cloud-init query instance-id

I wouldn't recommend this. If people still need instance_id, I think jinja is the way to go.

Reply:
I think jinja templates is effective enough to provide the flexibility to template our content based on cloud-init config details/values. We don't necessarily have to have environment variables..... And in some cases we don't want the security exposure of passing all cloud-init instance data as env vars to subprocess as I believe they can be inspected on the running process so it's probably generally a bad idea to expose potentially sensitive information as env vars to a subprocess calls anyway. So, I'm good with jinja being our primary means for providing this flexibility in user-data. thanks!

Reply:
> I think this is good enough. Warnings rising from the code are nice to have, but I don't think they're must have. 

@TheRealFalcon Just curious, why not? Why only include some things in docs and some things in code? These two mechanisms notify users in different ways. A user that has "set it and forget it" for their infrastructure may not be reading the docs. I'm not sure I follow the motivation for diverging.

Reply:
> And in some cases we don't want the security exposure of passing all cloud-init instance data as env vars to subprocess as I believe they can be inspected on the running process so it's probably generally a bad idea to expose potentially sensitive information as env vars to a subprocess calls anyway.

@blackboxsw Cloud-init doesn't do anything to filter environment variables from subprocesses; are you suggesting that's something that cloud-init should do? Environment variables are inherited by child processes, but cloud-init and its child processes run as root so what does it matter? 

> So, I'm good with jinja being our primary means for providing this flexibility in user-data. thanks!

+1 agreed

Reply:
> do anything to filter environment variables from subprocesses; are you suggesting that's something that cloud-init should do?

No, I was just thinking if we were to grow the environment variables we set for subp calls it may be worth avoiding keys which may contain sensitive values. But, per your second comment below. this is likely irrelevant anyway as cloud-init runs and spawns processes root and only root users will have perms to see those env variables anyway.

> but cloud-init and its child processes run as root so what does it matter?

Yeah this element doesn't really matter as /proc/<PID>/environ will only be accessible to root, so it won't be visible to non-admin/non-root users.

Reply:
> @TheRealFalcon Just curious, why not? Why only include some things in docs and some things in code? These two mechanisms notify users in different ways. A user that has "set it and forget it" for their infrastructure may not be reading the docs. I'm not sure I follow the motivation for diverging.

Because this requires grepping user scripts and that feels too invasive and error prone. If there's an easy way to support it in code, we should, but I don't want to commit atrocities just to give people a heads up that we might some day break something. There have been plenty of times we've done it with no warning at all (ie our breaking changes page), so I think a docs heads up is still more than enough.